### PR TITLE
Parallel Suffix cleanup

### PR DESCRIPTION
**Pull request by jamierpond**
- **wip**
- **think this is the best**

### CODE REVIEW

File: inc/zoo/swar/associative_iteration.h [hunk 1]
```diff
@@ -41,51 +41,26 @@ std::ostream &operator<<(std::ostream &out, zoo::swar::SWAR<NB, B> s) {
...
-        // 001...1
-        // 00001...1
-        // 000000001...1
-        shiftClearingMask =
-            shiftClearingMask &
-                S{static_cast<B>(shiftClearingMask.value() >> power)};
-        ZTE(power << 1);
+    for (;;) {
+        result = result ^ result.shiftIntraLaneLeft(power, shiftMask);
+        if (!--log2Count) { break; }
```
Comment (jamierpond) [inline inc/zoo/swar/associative_iteration.h:59]: @thecppzoo note one condition in the inner loop now

File: inc/zoo/swar/associative_iteration.h [hunk 2]
```diff
@@ -409,6 +422,74 @@ constexpr auto associativeOperatorIterated_regressive(
...
+static_assert(ps<uint32_t>(0b00000000'00000111) == 0b11111111'11111111'11111111'11111101);
+static_assert(ps<uint32_t>(0b00000000'00001111) == 0b00000000'00000000'00000000'00000101);
+static_assert(ps<uint32_t>(0b00000000'00011111) == 0b11111111'11111111'11111111'11110101);
+static_assert(ps<uint32_t>(0b00000000'00111111) == 0b00000000'00000000'00000000'00010101);
+static_assert(ps<uint32_t>(0b00000000'01111111) == 0b11111111'11111111'11111111'11010101);
+static_assert(ps<uint32_t>(0b00000000'11111111) == 0b00000000'00000000'00000000'01010101);
+
+template<typename S>
+constexpr auto parallel_suffix(S input) {
+    constexpr auto log2Count = S::Lanes;
```
Comment (thecppzoo) [inline inc/zoo/swar/associative_iteration.h]: This can't be right, the parallel suffix does not depend on the number of lanes, but the number of bits in the lanes

File: inc/zoo/swar/associative_iteration.h (hunk 1 above)
Comment (thecppzoo) [reply inc/zoo/swar/associative_iteration.h:59]: This condition is broken: it does not work with a bit count of non-powers of 2, like 7

File: inc/zoo/swar/associative_iteration.h (hunk 2 above)
Comment (jamierpond) [reply inc/zoo/swar/associative_iteration.h]: looks like you might be reviewing an outdated version ?

**Comment by thecppzoo** [comment]
This implementation might be simple enough, sure, but it can only accept lane sizes that have a power of two number of bits.
Let's review if the implementation I made is less efficient than yours.
Otherwise, the much harder challenge of supporting any arbitrary bitcount will have to decompose the number of bits into its binary representation to make the groups, and then AI would come to bear more clearly.
In simpler terms, this implementation is like multiplication when the factor is a power of two, much easier.

**Comment by jamierpond** [comment]
@thecppzoo https://godbolt.org/z/TEP36fhzh

**Comment by jamierpond** [comment]
hmmm... yeah i see what you mean...

**Comment by jamierpond** [comment]
ok now working with non-power of two nuim bits, just needed to idiot check
myself about the log2 impl for non-powers of two:
https://godbolt.org/z/aPa6q8r8c
